crates/router/src/core/api_keys.rs

```diff
@@ -0,0 +1,220 @@
+use common_utils::{date_time, errors::CustomResult, fp_utils};
+use error_stack::{report, IntoReport, ResultExt};
+use masking::{PeekInterface, Secret};
+use router_env::{instrument, tracing};
+
+use crate::{
+    consts,
+    core::errors::{self, RouterResponse, StorageErrorExt},
+    db::StorageInterface,
+    services::ApplicationResponse,
+    types::{api, storage, transformers::ForeignInto},
+    utils,
+};
+
+// Defining new types `PlaintextApiKey` and `HashedApiKey` in the hopes of reducing the possibility
+// of plaintext API key being stored in the data store.
+pub struct PlaintextApiKey(Secret<String>);
+pub struct HashedApiKey(String);
+
+impl PlaintextApiKey {
+    const HASH_KEY_LEN: usize = 32;
+
+    const PREFIX_LEN: usize = 12;
+
+    pub fn new(length: usize) -> Self {
```

Comment:
is key generation based on the environment in which its running? Or should this be pased? Or should we always generate a snd / prd key pair?

Reply:
It is based on the environment in which the app is running. The information is obtained from the `router_env` crate:

https://github.com/juspay/hyperswitch/blob/8205b2278d5a745d047274f17798a55e4c790a7f/crates/router_env/src/env.rs#L31-L38